File: LeetCodeProblemsCMake/Problems/LeetCode2410MaximumMatchingOfPlayersWithTrainers.cpp
//
// Created by saeed on 7/13/25.
//

#include "LeetCode2410MaximumMatchingOfPlayersWithTrainers.h"

#include <algorithm>
using namespace std;
int LeetCode2410MaximumMatchingOfPlayersWithTrainers::matchPlayersAndTrainers( vector<int> &players,  vector<int> &trainers) {
    ranges::sort(trainers);
    ranges::sort(players);
    int result{};
    int playersPointer{};
    int trainersPointer{};
    int playersLength= players.size();
    int trainersLength = trainers.size();
    while (playersPointer < playersLength && trainersPointer < trainersLength) {
        if (players[playersPointer] > trainers[trainersPointer]) {
            trainersPointer++;
            continue;
        }
        playersPointer++;
        trainersPointer++;
        result++;
    }
    return  result;
}

File: LeetCodeProblemsGo/LeetCode2410MaximumMatchingOfPlayersWithTrainers/LeetCode2410MaximumMatchingOfPlayersWithTrainers.go
package LeetCode2410MaximumMatchingOfPlayersWithTrainers

import "sort"

func matchPlayersAndTrainers(players []int, trainers []int) int {
	sort.Slice(players, func(i, j int) bool {
		return players[i] < players[j]
	})
	sort.Slice(trainers, func(i, j int) bool {
		return trainers[i] < trainers[j]
	})
	playersLength := len(players)
	playerPointer := 0
	trainersLength := len(trainers)
	trainerPointer := 0
	result := 0
	for playerPointer < playersLength && trainerPointer < trainersLength {
		if players[playerPointer] > trainers[trainerPointer] {
			trainerPointer++
			continue
		}
		playerPointer++
		trainerPointer++
		result++
	}
	return result
}
